README.md
# MX Range Slider plugin for Craft CMS 3.x

Is an easy, flexible and responsive range slider for Craft CMS with tons of options.

![Screenshot](resources/img/mx-rangeslider-craft.gif)

## Requirements

This plugin requires Craft CMS 3.0.0-beta.23 or later.

## Installation

1. Install with Composer via `composer require maxlazar/mx-range-slider` from your project directory
2. Install plugin in the Craft Control Panel under Settings > Plugins

You can also install MX Range Slider via the **Plugin Store** in the Craft Control Panel.

## MX Range Slider Overview



## Configuring MX Range Slider


| Option | Data-Attr | Defaults | Type | Description |
| --- | --- | --- | --- | --- |
| `skin` | `data-skin` | `flat` | string | Choose UI skin to use (flat, big, modern, round, sharp, square) |
| `type` | `data-type` | `single` | string | Choose slider type, could be `single` - for one handle, or `double` for two handles |
| `min` | `data-min` | `10` | number | Set slider minimum value |
| `max` | `data-max` | `100` | number | Set slider maximum value |
| `from` | `data-from` | `min` | number | Set start position for left handle (or for single handle) |
| `to` | `data-to` | `max` | number | Set start position for right handle |
| `step` | `data-step` | `1` | number | Set sliders step. Always > 0. Could be fractional |
| `min_interval` | `data-min-interval` | `-` | number | Set minimum diapason between sliders. Only for **double** type |
| `max_interval` | `data-max-interval` | `-` | number | Set minimum maximum between sliders. Only for **double** type |
| `drag_interval` | `data-drag-interval` | `false` | boolean | Allow user to drag whole range. Only for **double** type |
| `values` | `data-values` | `[]` | array | Set up your own array of possible slider values. They could be numbers or strings. If the values array is set up, min, max and step param, can no longer be changed |
| `from_fixed` | `data-from-fixed` | `false` | boolean | Fix position of left (or single) handle |
| `from_min` | `data-from-min` | `min` | number | Set minimum limit for left (or single) handle |
| `from_max` | `data-from-max` | `max` | number | Set maximum limit for left (or single) handle |
| `from_shadow` | `data-from-shadow` | `false` | boolean | Highlight the limits for left handle |
| `to_fixed` | `data-to-fixed` | `false` | boolean | Fix position of right handle |
| `to_min` | `data-to-min` | `min` | number | Set minimum limit for right handle |
| `to_max` | `data-to-max` | `max` | number | Set maximum limit for right handle |
| `to_shadow` | `data-to-shadow` | `false` | boolean | Highlight the right handle |
| `prettify_enabled` | `data-prettify-enabled` | `true` | boolean | Improve readability of long numbers: 10000000 &rarr; 10 000 000 |
| `prettify_separator` | `data-prettify-separator` | ` ` | string | Set up your own separator for long numbers: 10000000 &rarr; 10,000,000 etc. |
| `prettify` | `-` | `null` | function | Set up your own prettify function. Can be anything. For example, you can set up unix time as slider values and than transform them to cool looking dates |
| `force_edges` | `data-force-edges` | `false` | boolean | Sliders handles and tooltips will be always inside it's container |
| `keyboard` | `data-keyboard` | `true` | boolean | Activates keyboard controls. Move left: &larr;, &darr;, A, S. Move right: &rarr;, &uarr;, W, D. |
| `grid` | `data-grid` | `true` | boolean | Enables grid of values above the slider |
| `grid_margin` | `data-grid-margin` | `true` | boolean | Set left and right grid gaps |
| `grid_num` | `data-grid-num` | `4` | number | Number of grid units |
| `grid_snap` | `data-grid-snap` | `false` | boolean | Snap grid to sliders step (step param). If activated, grid_num will not be used. Max steps = 50 |
| `hide_min_max` | `data-hide-min-max` | `false` | boolean | Hides **min** and **max** labels |
| `hide_from_to` | `data-hide-from-to` | `false` | boolean | Hides **from** and **to** labels |
| `prefix` | `data-prefix` | `` | string | Set prefix for values. Will be set up right before the number: **$**100 |
| `postfix` | `data-postfix` | `` | string | Set postfix for values. Will be set up right after the number: 100**k** |
| `max_postfix` | `data-max-postfix` | `` | string | Special postfix, used only for maximum value. Will be showed after handle will reach maximum right position. For example **0 — 100+** |
| `decorate_both` | `data-decorate-both` | `true` | boolean | Used for **double** type and only if prefix or postfix was set up. Determine how to decorate close values. For example: **$10k — $100k** or **$10 — 100k** |
| `values_separator` | `data-decorate-both` | ` - ` | string | Set your own separator for close values. Used for **double** type. Default: **10 — 100**. Or you may set: **10 to 100, 10 + 100, 10 &rarr; 100** etc. |
| `disable` | `data-disable` | `false` | boolean | Locks slider and makes it inactive. Input is disabled too. Invisible to forms |
| `block` | `data-blokc` | `false` | boolean | Locks slider and makes it inactive. Input is NOT disabled. Can be send with forms |
| `extra_classes` | `data-extra-classes` | `—` | string | Traverse extra CSS-classes to sliders container |
| `scope` | `-` | `null` | object | Scope for callbacks. Pass any object |
| `onStart` | `-` | `null` | function | Callback. Is called on slider start. Gets all slider data as a 1st attribute |
| `onChange` | `-` | `null` | function | Callback. IS called on each values change. Gets all slider data as a 1st attribute |
| `onFinish` | `-` | `null` | function | Callback. Is called when user releases handle. Gets all slider data as a 1st attribute |
| `onUpdate` | `-` | `null` | function | Callback. Is called when slider is modified by external methods `update` or `reset` |


## Using the MX Range Slider plugin in your templates

    	from {{ entry.range_field.from }}
    	to {{ entry.range_field.to }}
    	value {{ entry.range_field.value }}

## MX Range Slider Roadmap


* Release it


### License

The MX RangeSlider for CraftCMS CE is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT)

### Thanks To

[Denis Ineshin](https://github.com/IonDen) for [ion.rangeSlider](https://github.com/IonDen/ion.rangeSlider) ([MIT license](http://opensource.org/licenses/MIT))


### And one more thing

Brought to you by [Max Lazar](https://www.wiseupstudio.com)



CHANGELOG.md
# MX Range Slider Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 3.1.0 - 2019-07-31
### Added
- Initial release


src/MxRangeSlider.php
<?php
/**
 * MX Range Slider plugin for Craft CMS 3.x
 *
 * Is an easy, flexible and responsive range slider for Craft CMS with tons of options.
 *
 * @link      https://www.wiseupstudio.com
 * @copyright Copyright (c) 2019 Max Lazar
 */

namespace maxlazar\mxrangeslider;

use maxlazar\mxrangeslider\variables\MxRangeSliderVariable;
use maxlazar\mxrangeslider\twigextensions\MxRangeSliderTwigExtension;
use maxlazar\mxrangeslider\fields\RangeSlider as RangeSliderField;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\services\Fields;
use craft\web\twig\variables\CraftVariable;
use craft\events\RegisterComponentTypesEvent;

use yii\base\Event;

/**
 * Craft plugins are very much like little applications in and of themselves. We’ve made
 * it as simple as we can, but the training wheels are off. A little prior knowledge is
 * going to be required to write a plugin.
 *
 * For the purposes of the plugin docs, we’re going to assume that you know PHP and SQL,
 * as well as some semi-advanced concepts like object-oriented programming and PHP namespaces.
 *
 * https://craftcms.com/docs/plugins/introduction
 *
 * @author    Max Lazar
 * @package   MxRangeSlider
 * @since     3.1.1
 *
 */
class MxRangeSlider extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * Static property that is an instance of this plugin class so that it can be accessed via
     * MxRangeSlider::$plugin
     *
     * @var MxRangeSlider
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * To execute your plugin’s migrations, you’ll need to increase its schema version.
     *
     * @var string
     */
    public $schemaVersion = '3.1.2';

    // Public Methods
    // =========================================================================

    /**
     * Set our $plugin static property to this class so that it can be accessed via
     * MxRangeSlider::$plugin
     *
     * Called after the plugin class is instantiated; do any one-time initialization
     * here such as hooks and events.
     *
     * If you have a '/vendor/autoload.php' file, it will be loaded for you automatically;
     * you do not need to load it in your init() method.
     *
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        // Add in our Twig extensions
        Craft::$app->view->registerTwigExtension(new MxRangeSliderTwigExtension());

        // Register our fields
        Event::on(
            Fields::class,
            Fields::EVENT_REGISTER_FIELD_TYPES,
            function (RegisterComponentTypesEvent $event) {
                $event->types[] = RangeSliderField::class;
            }
        );

        // Register our variables
        Event::on(
            CraftVariable::class,
            CraftVariable::EVENT_INIT,
            function (Event $event) {
                /** @var CraftVariable $variable */
                $variable = $event->sender;
                $variable->set('mxRangeSlider', MxRangeSliderVariable::class);
            }
        );

        // Do something after we're installed
        Event::on(
            Plugins::class,
            Plugins::EVENT_AFTER_INSTALL_PLUGIN,
            function (PluginEvent $event) {
                if ($event->plugin === $this) {
                    // We were just installed
                }
            }
        );

/**
 * Logging in Craft involves using one of the following methods:
 *
 * Craft::trace(): record a message to trace how a piece of code runs. This is mainly for development use.
 * Craft::info(): record a message that conveys some useful information.
 * Craft::warning(): record a warning message that indicates something unexpected has happened.
 * Craft::error(): record a fatal error that should be investigated as soon as possible.
 *
 * Unless `devMode` is on, only Craft::warning() & Craft::error() will log to `craft/storage/logs/web.log`
 *
 * It's recommended that you pass in the magic constant `__METHOD__` as the second parameter, which sets
 * the category to the method (prefixed with the fully qualified class name) where the constant appears.
 *
 * To enable the Yii debug toolbar, go to your user account in the AdminCP and check the
 * [] Show the debug toolbar on the front end & [] Show the debug toolbar on the Control Panel
 *
 * http://www.yiiframework.com/doc-2.0/guide-runtime-logging.html
 */
        Craft::info(
            Craft::t(
                'mx-range-slider',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

    // Protected Methods
    // =========================================================================
}


src/helpers/Config.php
<?php
/**
 * MX Range Slider plugin for Craft CMS 3.x.
 *
 * @see      https://www.wiseupstudio.com
 *
 * @copyright Copyright (c) 2019 MaxLazar
 */

namespace maxlazar\mxrangeslider\helpers;

use Craft;
use craft\helpers\StringHelper;

/**
 * @author    Max Lazar
 *
 * @since     3.1.0
 */
class Config
{
    // Constants
    // =========================================================================
    const LOCAL_CONFIG_DIR = 'settings';

    // Static Methods
    // =========================================================================

    /**
     * Loads a config file from, trying @craft/config first, then the plugin's.
     *
     *
     * @param string $filePath
     * @param null   $alias
     *
     * @return array
     */
    public static function getConfigFromFile(string $filePath, $alias = null): array
    {
        // Try craft/config first
        $path = self::getConfigFilePath('@config', $filePath);
        if (!file_exists($path)) {
            // Now try our own internal config
            $path = self::getConfigFilePath('@maxlazar/mxrangeslider', $filePath);
            if (!file_exists($path)) {
                if (!$alias) {
                    return [];
                }
                // Now the additional alias config
                $path = self::getConfigFilePath($alias, $filePath);
                if (!file_exists($path)) {
                    return [];
                }
            }
        }

        if (!\is_array($config = @include $path)) {
            return [];
        }

        // If it's not a multi-environment config, return the whole thing
        if (!array_key_exists('*', $config)) {
            return $config;
        }

    }

    /**
     * Load an array of config files, merging them together in a single array.
     *
     * @param array $filePaths
     *
     * @return array
     */
    public static function getMergedConfigFromFiles(array $filePaths): array
    {
        $mergedConfig = [];
        foreach ($filePaths as $filePath) {
            $mergedConfig = ArrayHelper::merge($mergedConfig, self::getConfigFromFile($filePath));
        }

        return $mergedConfig;
    }

    // Private Methods
    // =========================================================================

    /**
     * Return a path from an alias and a partial path.
     *
     * @param string $alias
     * @param string $filePath
     *
     * @return string
     */
    private static function getConfigFilePath(string $alias, string $filePath): string
    {
        $path = DIRECTORY_SEPARATOR.ltrim($filePath, DIRECTORY_SEPARATOR);
        $path = Craft::getAlias($alias)
            .DIRECTORY_SEPARATOR
            .self::LOCAL_CONFIG_DIR
            .str_replace(array('/', '\\'), DIRECTORY_SEPARATOR, $path)
            .'.php';

        return $path;
    }
}


src/fields/RangeSlider.php
<?php
/**
 * MX Range Slider plugin for Craft CMS 3.x.
 *
 * Is an easy, flexible and responsive range slider for Craft CMS with tons of options.
 *
 * @see      https://www.wiseupstudio.com
 *
 * @copyright Copyright (c) 2019 Max Lazar
 */

namespace maxlazar\mxrangeslider\fields;

use maxlazar\mxrangeslider\assetbundles\rangesliderfield\RangeSliderFieldAsset;
use maxlazar\mxrangeslider\helpers\Config as ConfigLoader;
use Craft;
use craft\base\ElementInterface;
use craft\base\Field;
use yii\db\Schema;
use craft\helpers\Json as JsonHelper;

/**
 * RangeSlider Field.
 *
 * Whenever someone creates a new field in Craft, they must specify what
 * type of field it is. The system comes with a handful of field types baked in,
 * and we’ve made it extremely easy for plugins to add new ones.
 *
 * https://craftcms.com/docs/plugins/field-types
 *
 * @author    Max Lazar
 *
 * @since     3.1.0
 */
class RangeSlider extends Field
{
    // Public Properties
    // =========================================================================

    /**
     * Some attribute.
     *
     * @var string
     */
    public $someAttribute = 'Some Default';
    public $options;

    // Static Methods
    // =========================================================================

    /**
     * Returns the display name of this class.
     *
     * @return string the display name of this class
     */
    public static function displayName(): string
    {
        return Craft::t('mx-range-slider', 'Range Slider');
    }

    // Public Methods
    // =========================================================================

    /**
     * Returns the validation rules for attributes.
     *
     * Validation rules are used by [[validate()]] to check if attribute values are valid.
     * Child classes may override this method to declare different validation rules.
     *
     * More info: http://www.yiiframework.com/doc-2.0/guide-input-validation.html
     *
     * @return array
     */
    public function rules()
    {
        $rules = parent::rules();
        $rules = array_merge($rules, [
        ]);

        return $rules;
    }

    /**
     * Returns the column type that this field should get within the content table.
     *
     * This method will only be called if [[hasContentColumn()]] returns true.
     *
     * @return string The column type. [[\yii\db\QueryBuilder::getColumnType()]] will be called
     *                to convert the give column type to the physical one. For example, `string` will be converted
     *                as `varchar(255)` and `string(100)` becomes `varchar(100)`. `not null` will automatically be
     *                appended as well.
     *
     * @see \yii\db\QueryBuilder::getColumnType()
     */
    public function getContentColumnType(): string
    {
        return Schema::TYPE_STRING;
    }

    /**
     * Normalizes the field’s value for use.
     *
     * This method is called when the field’s value is first accessed from the element. For example, the first time
     * `entry.myFieldHandle` is called from a template, or right before [[getInputHtml()]] is called. Whatever
     * this method returns is what `entry.myFieldHandle` will likewise return, and what [[getInputHtml()]]’s and
     * [[serializeValue()]]’s $value arguments will be set to.
     *
     * @param mixed                 $value   The raw field value
     * @param ElementInterface|null $element The element the field is associated with, if there is one
     *
     * @return mixed The prepared field value
     */
    public function normalizeValue($value, ElementInterface $element = null)
    {
        $data = array();

        if (is_string($value)) {
            $value = JsonHelper::decodeIfJson($value);
        }

        if (isset($value['from'])) {
            return $value;
        }

        $minmax = explode(';', $value);

        $minmax[1] = (!empty($value) && count($minmax) > 1) ? $minmax[1] : $minmax[0];
        $data['from'] = $minmax[0];
        $data['to'] = $minmax[1];
        $data['value'] = $minmax[0];

        return $data;
    }

    /**
     * Returns the component’s settings HTML.
     *
     * @return string|null
     */
    public function getSettingsHtml()
    {
        $FieldsConfig = ConfigLoader::getConfigFromFile('RangeSliderField');

        // Render the settings template
        return Craft::$app->getView()->renderTemplate(
            'mx-range-slider/_components/fields/RangeSlider_settings',
            [
                'field' => $this,
                'FieldsConfig' => $FieldsConfig,
            ]
        );
    }

    /**
     * Returns the field’s input HTML.
     *
     * @param mixed                 $value   The field’s value. This will either be the [[normalizeValue() normalized value]],
     *                                       raw POST data (i.e. if there was a validation error), or null
     * @param ElementInterface|null $element The element the field is associated with, if there is one
     *
     * @return string the input HTML
     */
    public function getInputHtml($value, ElementInterface $element = null): string
    {
        $variables = [];

        // Register our asset bundle
        Craft::$app->getView()->registerAssetBundle(RangeSliderFieldAsset::class);

        // Basic variables
        $variables['name'] = $this->handle;
        $variables['value'] = $value['from'].(('double' == $this->options['type']) ? ';'.$value['to'] : '');
        $variables['field'] = $this;

        // Get our id and namespace
        $id = Craft::$app->getView()->formatInputId($this->handle);
        $nameSpacedId = Craft::$app->getView()->namespaceInputId($id);
        $variables['id'] = $id;
        $variables['nameSpacedId'] = $nameSpacedId;

        $FieldsConfig = ConfigLoader::getConfigFromFile('RangeSliderField');

        Craft::$app->getView()->registerJs("var {$id}_cnf = {".$this->buildOptions($FieldsConfig)."};
        setTimeout(function() { $('#".$nameSpacedId."').ionRangeSlider( {$id}_cnf )}, 150);");

        // Render the input template
        return Craft::$app->getView()->renderTemplate(
            'mx-range-slider/_components/fields/RangeSlider_input',
            $variables
        );
    }

    public function buildOptions($data)
    {
        $js = '';

        foreach ($data as $key => $value) {
            if (!isset($this->options[$key]) || $value['defaults'] == $this->options[$key]) {
                continue;
            }
            if ('string' == $value['type'] || 'number' == $value['type'] || 'select' == $value['type']) {
                $js .= $key.': "'.$this->options[$key].'",'."\n";
            }
            if ('boolean' == $value['type']) {
                $js .= $key.': '.('' == $this->options[$key] ? 'false' : 'true').', '."\n";
            }
            if ('function' == $value['type']) {
                $js .= $key.': function(data){'."\n".$this->options[$key]."\n".'},'."\n";
            }
        }

        return $js;
    }
}


src/variables/MxRangeSliderVariable.php
<?php
/**
 * MX Range Slider plugin for Craft CMS 3.x
 *
 * Is an easy, flexible and responsive range slider for Craft CMS with tons of options.
 *
 * @link      https://www.wiseupstudio.com
 * @copyright Copyright (c) 2019 Max Lazar
 */

namespace maxlazar\mxrangeslider\variables;

use maxlazar\mxrangeslider\MxRangeSlider;

use Craft;

/**
 * MX Range Slider Variable
 *
 * Craft allows plugins to provide their own template variables, accessible from
 * the {{ craft }} global variable (e.g. {{ craft.mxRangeSlider }}).
 *
 * https://craftcms.com/docs/plugins/variables
 *
 * @author    Max Lazar
 * @package   MxRangeSlider
 * @since     3.1.0
 */
class MxRangeSliderVariable
{
    // Public Methods
    // =========================================================================

    /**
     * Whatever you want to output to a Twig template can go into a Variable method.
     * You can have as many variable functions as you want.  From any Twig template,
     * call it like this:
     *
     *     {{ craft.mxRangeSlider.exampleVariable }}
     *
     * Or, if your variable requires parameters from Twig:
     *
     *     {{ craft.mxRangeSlider.exampleVariable(twigValue) }}
     *
     * @param null $optional
     * @return string
     */
    public function exampleVariable($optional = null)
    {
        $result = "And away we go to the Twig template...";
        if ($optional) {
            $result = "I'm feeling optional today...";
        }
        return $result;
    }
}


src/twigextensions/MxRangeSliderTwigExtension.php
<?php
/**
 * MX Range Slider plugin for Craft CMS 3.x
 *
 * Is an easy, flexible and responsive range slider for Craft CMS with tons of options.
 *
 * @link      https://www.wiseupstudio.com
 * @copyright Copyright (c) 2019 Max Lazar
 */

namespace maxlazar\mxrangeslider\twigextensions;

use maxlazar\mxrangeslider\MxRangeSlider;

use Craft;

/**
 * Twig can be extended in many ways; you can add extra tags, filters, tests, operators,
 * global variables, and functions. You can even extend the parser itself with
 * node visitors.
 *
 * http://twig.sensiolabs.org/doc/advanced.html
 *
 * @author    Max Lazar
 * @package   MxRangeSlider
 * @since     3.1.0
 */
class MxRangeSliderTwigExtension extends \Twig_Extension
{
    // Public Methods
    // =========================================================================

    /**
     * Returns the name of the extension.
     *
     * @return string The extension name
     */
    public function getName()
    {
        return 'MxRangeSlider';
    }

    /**
     * Returns an array of Twig filters, used in Twig templates via:
     *
     *      {{ 'something' | someFilter }}
     *
     * @return array
     */
    public function getFilters()
    {
        return [
            new \Twig_SimpleFilter('someFilter', [$this, 'someInternalFunction']),
        ];
    }

    /**
     * Returns an array of Twig functions, used in Twig templates via:
     *
     *      {% set this = someFunction('something') %}
     *
    * @return array
     */
    public function getFunctions()
    {
        return [
            new \Twig_SimpleFunction('someFunction', [$this, 'someInternalFunction']),
        ];
    }

    /**
     * Our function called via Twig; it can do anything you want
     *
     * @param null $text
     *
     * @return string
     */
    public function someInternalFunction($text = null)
    {
        $result = $text . " in the way";

        return $result;
    }
}


src/settings/RangeSliderField.php
<?php

/**
 * MX Range Slider plugin for Craft CMS 3.x.
 *
 *
 * @see      https://www.wiseupstudio.com
 *
 * @copyright Copyright (c) 2019 maxlazar
 */

/**
 * @author    maxlazar
 *
 * @since     3.0.0
 */

return [
  'skin' => [
    'option' => 'skin',
    'data-attr' => 'data-skin',
    'defaults' => 'modern',
    'type' => 'select',
    'values' => [
      'flat' => 'flat',
      'big' => 'big',
      'modern' => 'modern',
      'round' => 'round',
      'sharp' => 'sharp',
      'square' => 'square',
    ],
    'description' => 'Choose UI skin to use (flat, big, modern, round, sharp, square)',
  ],
  'type' => [
    'option' => 'type',
    'data-attr' => 'data-type',
    'defaults' => 'single',
    'type' => 'select',
    'values' => [
      'single' => 'single',
      'double' => 'double',
    ],
    'description' => 'Choose slider type, could be single - for one handle, or double for two handles',
  ],
  'min' => [
    'option' => 'min',
    'data-attr' => 'data-min',
    'defaults' => 10,
    'type' => 'number',
    'description' => 'Set slider minimum value',
  ],
  'max' => [
    'option' => 'max',
    'data-attr' => 'data-max',
    'defaults' => 100,
    'type' => 'number',
    'description' => 'Set slider maximum value',
  ],
  'from' => [
    'option' => 'from',
    'data-attr' => 'data-from',
    'defaults' => 'min',
    'type' => 'number',
    'description' => 'Set start position for left handle (or for single handle)',
  ],
  'to' => [
    'option' => 'to',
    'data-attr' => 'data-to',
    'defaults' => 'max',
    'type' => 'number',
    'description' => 'Set start position for right handle',
  ],
  'step' => [
    'option' => 'step',
    'data-attr' => 'data-step',
    'defaults' => 1,
    'type' => 'number',
    'description' => 'Set sliders step. Always > 0. Could be fractional',
  ],
  'min_interval' => [
    'option' => 'min_interval',
    'data-attr' => 'data-min-interval',
    'defaults' => '-',
    'type' => 'number',
    'description' => 'Set minimum diapason between sliders. Only for double type',
  ],
  'max_interval' => [
    'option' => 'max_interval',
    'data-attr' => 'data-max-interval',
    'defaults' => '-',
    'type' => 'number',
    'description' => 'Set minimum maximum between sliders. Only for double type',
  ],
  'drag_interval' => [
    'option' => 'drag_interval',
    'data-attr' => 'data-drag-interval',
    'defaults' => '0',
    'type' => 'boolean',
    'description' => 'Allow user to drag whole range. Only for double type',
  ],
  'values' => [
    'option' => 'values',
    'data-attr' => 'data-values',
    'defaults' => [],
    'type' => 'array',
    'description' => 'Set up your own array of possible slider values. They could be numbers or strings. If the values array is set up, min, max and step param, can no longer be changed',
  ],
  'from_fixed' => [
    'option' => 'from_fixed',
    'data-attr' => 'data-from-fixed',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Fix position of left (or single) handle',
  ],
  'from_min' => [
    'option' => 'from_min',
    'data-attr' => 'data-from-min',
    'defaults' => 'min',
    'type' => 'number',
    'description' => 'Set minimum limit for left (or single) handle',
  ],
  'from_max' => [
    'option' => 'from_max',
    'data-attr' => 'data-from-max',
    'defaults' => 'max',
    'type' => 'number',
    'description' => 'Set maximum limit for left (or single) handle',
  ],
  'from_shadow' => [
    'option' => 'from_shadow',
    'data-attr' => 'data-from-shadow',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Highlight the limits for left handle',
  ],
  'to_fixed' => [
    'option' => 'to_fixed',
    'data-attr' => 'data-to-fixed',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Fix position of right handle',
  ],
  'to_min' => [
    'option' => 'to_min',
    'data-attr' => 'data-to-min',
    'defaults' => 'min',
    'type' => 'number',
    'description' => 'Set minimum limit for right handle',
  ],
  'to_max' => [
    'option' => 'to_max',
    'data-attr' => 'data-to-max',
    'defaults' => 'max',
    'type' => 'number',
    'description' => 'Set maximum limit for right handle',
  ],
  'to_shadow' => [
    'option' => 'to_shadow',
    'data-attr' => 'data-to-shadow',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Highlight the right handle',
  ],
  'prettify_enabled' => [
    'option' => 'prettify_enabled',
    'data-attr' => 'data-prettify-enabled',
    'defaults' => 1,
    'type' => 'boolean',
    'description' => 'Improve readability of long numbers: 10000000 → 10 000 000',
  ],
  'prettify_separator' => [
    'option' => 'prettify_separator',
    'data-attr' => 'data-prettify-separator',
    'defaults' => '',
    'type' => 'string',
    'description' => 'Set up your own separator for long numbers: 10000000 → 10,000,000 etc.',
  ],
  'force_edges' => [
    'option' => 'force_edges',
    'data-attr' => 'data-force-edges',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Sliders handles and tooltips will be always inside it\'s container',
  ],
  'keyboard' => [
    'option' => 'keyboard',
    'data-attr' => 'data-keyboard',
    'defaults' => 1,
    'type' => 'boolean',
    'description' => 'Activates keyboard controls. Move left: ←, ↓, A, S. Move right: →, ↑, W, D.',
  ],
  'grid' => [
    'option' => 'grid',
    'data-attr' => 'data-grid',
    'defaults' => 1,
    'type' => 'boolean',
    'description' => 'Enables grid of values above the slider',
  ],
  'grid_margin' => [
    'option' => 'grid_margin',
    'data-attr' => 'data-grid-margin',
    'defaults' => 1,
    'type' => 'boolean',
    'description' => 'Set left and right grid gaps',
  ],
  'grid_num' => [
    'option' => 'grid_num',
    'data-attr' => 'data-grid-num',
    'defaults' => 4,
    'type' => 'number',
    'description' => 'Number of grid units',
  ],
  'grid_snap' => [
    'option' => 'grid_snap',
    'data-attr' => 'data-grid-snap',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Snap grid to sliders step (step param). If activated, grid_num will not be used. Max steps = 50',
  ],
  'hide_min_max' => [
    'option' => 'hide_min_max',
    'data-attr' => 'data-hide-min-max',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Hides min and max labels',
  ],
  'hide_from_to' => [
    'option' => 'hide_from_to',
    'data-attr' => 'data-hide-from-to',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Hides from and to labels',
  ],
  'prefix' => [
    'option' => 'prefix',
    'data-attr' => 'data-prefix',
    'defaults' => '``',
    'type' => 'string',
    'description' => 'Set prefix for values. Will be set up right before the number: **$**100',
  ],
  'postfix' => [
    'option' => 'postfix',
    'data-attr' => 'data-postfix',
    'defaults' => '``',
    'type' => 'string',
    'description' => 'Set postfix for values. Will be set up right after the number: 100k',
  ],
  'max_postfix' => [
    'option' => 'max_postfix',
    'data-attr' => 'data-max-postfix',
    'defaults' => '``',
    'type' => 'string',
    'description' => 'Special postfix, used only for maximum value. Will be showed after handle will reach maximum right position. For example 0 — 100+',
  ],
  'decorate_both' => [
    'option' => 'decorate_both',
    'data-attr' => 'data-decorate-both',
    'defaults' => 1,
    'type' => 'boolean',
    'description' => 'Used for double type and only if prefix or postfix was set up. Determine how to decorate close values. For example: $10k — $100k or $10 — 100k',
  ],
  'values_separator' => [
    'option' => 'values_separator',
    'data-attr' => 'data-decorate-both',
    'defaults' => '-',
    'type' => 'string',
    'description' => 'Set your own separator for close values. Used for double type. Default: 10 — 100. Or you may set: 10 to 100, 10 + 100, 10 → 100 etc.',
  ],
  'disable' => [
    'option' => 'disable',
    'data-attr' => 'data-disable',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Locks slider and makes it inactive. Input is disabled too. Invisible to forms',
  ],
  'block' => [
    'option' => 'block',
    'data-attr' => 'data-blokc',
    'defaults' => 0,
    'type' => 'boolean',
    'description' => 'Locks slider and makes it inactive. Input is NOT disabled. Can be send with forms',
  ],
  'extra_classes' => [
    'option' => 'extra_classes',
    'data-attr' => 'data-extra-classes',
    'defaults' => '—',
    'type' => 'string',
    'description' => 'Traverse extra CSS-classes to sliders container',
  ],
  'scope' => [
    'option' => 'scope',
    'data-attr' => '-',
    'defaults' => null,
    'type' => 'object',
    'description' => 'Scope for callbacks. Pass any object',
  ],
  'prettify' => [
    'option' => 'prettify',
    'data-attr' => '-',
    'defaults' => null,
    'type' => 'function',
    'description' => 'Set up your own prettify function. Can be anything. For example, you can set up unix time as slider values and than transform them to cool looking dates',
  ],
  'onStart' => [
    'option' => 'onStart',
    'data-attr' => '-',
    'defaults' => null,
    'type' => 'function',
    'description' => 'Callback. Is called on slider start. Gets all slider data as a 1st attribute',
  ],
  'onChange' => [
    'option' => 'onChange',
    'data-attr' => '-',
    'defaults' => null,
    'type' => 'function',
    'description' => 'Callback. IS called on each values change. Gets all slider data as a 1st attribute',
  ],
  'onFinish' => [
    'option' => 'onFinish',
    'data-attr' => '-',
    'defaults' => null,
    'type' => 'function',
    'description' => 'Callback. Is called when user releases handle. Gets all slider data as a 1st attribute',
  ],
  'onUpdate' => [
    'option' => 'onUpdate',
    'data-attr' => '-',
    'defaults' => null,
    'type' => 'function',
    'description' => 'Callback. Is called when slider is modified by external methods update or reset',
  ],
];
